File: dify-agent/src/shellctl/server/tmux.py
"""tmux control layer for shellctl jobs.

The rest of shellctl treats this module as the only place that knows tmux CLI
 command shapes. Tests can replace `TmuxControllerProtocol` with a fake to
 exercise SQLite/output semantics without depending on a local tmux daemon.
"""

from __future__ import annotations

import os
import shlex
import subprocess
import tempfile
from pathlib import Path
from typing import Protocol, cast

import anyio

from shellctl.server.artifacts import (
    pipe_drained_path,
    pipe_error_log_path,
    pipe_failed_path,
    runner_ended_at_path,
    runner_exit_code_path,
)
from shellctl.server.config import ShellctlConfig
from shellctl.server.errors import ShellctlServerError
from shellctl.shared.runtime import (
    job_pane_target,
    job_session_name,
)
from shellctl.shared.schemas import TerminalSize


class TmuxControllerProtocol(Protocol):
    """Protocol used by `ShellctlService` for tmux interactions."""

    async def start_server(self) -> None: ...

    async def list_sessions(self) -> set[str]: ...

    async def session_exists(self, session_name: str) -> bool: ...

    async def is_output_pipe_active(self, *, job_id: str) -> bool | None: ...

    async def create_job_session(
        self,
        *,
        job_id: str,
        job_dir: Path,
        cwd: Path,
        terminal: TerminalSize,
    ) -> None: ...

    async def enable_output_pipe(
        self, *, job_id: str, job_dir: Path, ready_file: Path
    ) -> None: ...

    async def send_input(self, *, job_id: str, text: str) -> None: ...

    async def send_interrupt(self, *, job_id: str) -> None: ...

    async def cleanup_session(self, *, job_id: str) -> None: ...


class TmuxController:
    """Best-effort wrapper around a dedicated tmux socket.

    The controller always clears `TMUX` from the child environment and always
    passes `-S <socket>` so shellctl sessions stay isolated from the user's
    default tmux server.
    """

    def __init__(self, config: ShellctlConfig) -> None:
        self._config = config

    async def start_server(self) -> None:
        await self._run_tmux("start-server")

    async def list_sessions(self) -> set[str]:
        result = await self._run_tmux(
            "list-sessions", "-F", "#{session_name}", check=False
        )
        if result.returncode != 0:
            stderr = result.stderr.decode("utf-8", errors="replace")
            if _tmux_target_missing(stderr):
                return set()
            raise ShellctlServerError(
                500, "tmux_error", stderr.strip() or "tmux list-sessions failed"
            )
        output = result.stdout.decode("utf-8", errors="replace")
        return {line.strip() for line in output.splitlines() if line.strip()}

    async def session_exists(self, session_name: str) -> bool:
        return session_name in await self.list_sessions()

    async def is_output_pipe_active(self, *, job_id: str) -> bool | None:
        result = await self._run_tmux(
            "display-message",
            "-p",
            "-t",
            job_pane_target(job_id),
            "#{pane_pipe}",
            check=False,
        )
        if result.returncode != 0:
            stderr = result.stderr.decode("utf-8", errors="replace")
            if _tmux_target_missing(stderr):
                return None
            raise ShellctlServerError(
                500,
                "tmux_error",
                stderr.strip() or f"Failed to inspect output pipe for {job_id}",
            )
        return result.stdout.decode("utf-8", errors="replace").strip() == "1"

    async def create_job_session(
        self,
        *,
        job_id: str,
        job_dir: Path,
        cwd: Path,
        terminal: TerminalSize,
    ) -> None:
        runner_command = self._shell_join(
            [
                str(self._config.runner_path),
                str(job_dir),
                job_id,
                str(cwd),
            ]
        )
        result = await self._run_tmux(
            "-f",
            "/dev/null",
            "new-session",
            "-d",
            "-s",
            job_session_name(job_id),
            "-x",
            str(terminal.cols),
            "-y",
            str(terminal.rows),
            runner_command,
            check=False,
        )
        if result.returncode != 0:
            raise ShellctlServerError(
                500,
                "tmux_new_session_failed",
                result.stderr.decode("utf-8", errors="replace").strip()
                or f"Failed to create tmux session for {job_id}",
            )

    async def enable_output_pipe(
        self, *, job_id: str, job_dir: Path, ready_file: Path
    ) -> None:
        output_command = self._pipe_command_source(
            job_id=job_id,
            job_dir=job_dir,
            ready_file=ready_file,
        )
        result = await self._run_tmux(
            "pipe-pane",
            "-o",
            "-t",
            job_pane_target(job_id),
            output_command,
            check=False,
        )
        if result.returncode != 0:
            raise ShellctlServerError(
                500,
                "pipe_failed",
                result.stderr.decode("utf-8", errors="replace").strip()
                or f"Failed to attach output pipe for {job_id}",
            )

    def _pipe_command_source(
        self, *, job_id: str, job_dir: Path, ready_file: Path
    ) -> str:
        """Build the tmux `pipe-pane` command that drains and finalizes output.

        For normal exits, the runner now records completion metadata into job
        artifacts and the pipe finalizer commits `runner-exit` only after
        the lightweight sanitizer reaches EOF and flushes `output.log`
        successfully. Sanitizer stderr is captured into `pipe-error.log` so
        startup timeouts can distinguish slow imports from subprocess crashes.
        If the follow-up `runner-exit` write fails, the drain marker remains in
        place and stderr is appended to the same log with an explicit status
        line. The pipe still exits with the sanitizer status so a drained job is
        not misclassified as `pipe_failed` before reconciliation can recover the
        SQLite write from the drained artifacts.
        """

        sanitize_command = self._shell_join(
            (
                *self._config.sanitize_pty_command,
                "--ready-file",
                str(ready_file),
            )
        )
        runner_exit_command = self._shell_join(
            (
                *self._config.runner_exit_command,
                "--state-dir",
                str(self._config.state_dir),
                "--job-id",
                job_id,
                "--sqlite-busy-timeout-ms",
                str(self._config.sqlite_busy_timeout_ms),
            )
        )
        output_path = shlex.quote(str(job_dir / "output.log"))
        drained_path = shlex.quote(str(pipe_drained_path(job_dir)))
        error_log_path = shlex.quote(str(pipe_error_log_path(job_dir)))
        failed_path = shlex.quote(str(pipe_failed_path(job_dir)))
        exit_code_path = shlex.quote(str(runner_exit_code_path(job_dir)))
        ended_at_path = shlex.quote(str(runner_ended_at_path(job_dir)))
        return " ; ".join(
            [
                f"{sanitize_command} >> {output_path} 2> {error_log_path}",
                "sanitize_status=$?",
                "runner_exit_status=0",
                (
                    'if [ "$sanitize_status" -eq 0 ]; then '
                    f": > {drained_path}; "
                    f"if [ -s {exit_code_path} ] && [ -s {ended_at_path} ]; then "
                    f'{runner_exit_command} --exit-code "$(cat {exit_code_path})" '
                    f'--ended-at "$(cat {ended_at_path})" 2>> {error_log_path}; '
                    "runner_exit_status=$?; "
                    'if [ "$runner_exit_status" -ne 0 ]; then '
                    f"printf 'runner-exit failed with status %s\\n' \"$runner_exit_status\" >> {error_log_path}; "
                    "fi; fi; "
                    f"else : > {failed_path}; fi"
                ),
                'if [ "$sanitize_status" -ne 0 ]; then exit "$sanitize_status"; fi',
                'exit "$sanitize_status"',
            ]
        )

    async def send_input(self, *, job_id: str, text: str) -> None:
        buffer_name = f"shellctl-in-{job_id}"
        runtime_dir = cast(Path, self._config.runtime_dir)
        runtime_dir.mkdir(parents=True, exist_ok=True)
        fd, tmp_name = tempfile.mkstemp(
            prefix=f"shellctl-input-{job_id}-", dir=runtime_dir
        )
        tmp_path = Path(tmp_name)
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as handle:
                handle.write(text)
            load_result = await self._run_tmux(
                "load-buffer",
                "-b",
                buffer_name,
                str(tmp_path),
                check=False,
            )
            if load_result.returncode != 0:
                stderr = load_result.stderr.decode("utf-8", errors="replace").strip()
                if _tmux_target_missing(stderr):
                    raise ShellctlServerError(
                        409,
                        "tmux_target_missing",
                        stderr or f"The tmux pane for {job_id} is no longer available",
                    )
                raise ShellctlServerError(
                    500,
                    "tmux_input_failed",
                    stderr or f"Failed to load input buffer for {job_id}",
                )
            paste_result = await self._run_tmux(
                "paste-buffer",
                "-t",
                job_pane_target(job_id),
                "-b",
                buffer_name,
                check=False,
            )
            if paste_result.returncode != 0:
                stderr = paste_result.stderr.decode("utf-8", errors="replace").strip()
                if _tmux_target_missing(stderr):
                    raise ShellctlServerError(
                        409,
                        "tmux_target_missing",
                        stderr or f"The tmux pane for {job_id} is no longer available",
                    )
                raise ShellctlServerError(
                    500,
                    "tmux_input_failed",
                    stderr or f"Failed to paste input buffer for {job_id}",
                )
        finally:
            await self._run_tmux("delete-buffer", "-b", buffer_name, check=False)
            tmp_path.unlink(missing_ok=True)

    async def send_interrupt(self, *, job_id: str) -> None:
        await self._run_tmux(
            "send-keys",
            "-t",
            job_pane_target(job_id),
            "C-c",
            check=False,
        )

    async def cleanup_session(self, *, job_id: str) -> None:
        await self._run_tmux(
            "kill-session",
            "-t",
            job_session_name(job_id),
            check=False,
        )

    async def _run_tmux(
        self, *args: str, check: bool = True
    ) -> subprocess.CompletedProcess[bytes]:
        env = dict(os.environ)
        env.pop("TMUX", None)
        try:
            result = await anyio.run_process(
                ["tmux", "-S", str(self._config.tmux_socket), *args],
                env=env,
                check=False,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
            )
        except FileNotFoundError as exc:
            raise ShellctlServerError(
                500, "tmux_not_installed", "tmux executable was not found"
            ) from exc
        if check and result.returncode != 0:
            raise ShellctlServerError(
                500,
                "tmux_error",
                result.stderr.decode("utf-8", errors="replace").strip()
                or "tmux command failed",
            )
        return result

    @staticmethod
    def _shell_join(parts: tuple[str, ...] | list[str]) -> str:
        return " ".join(shlex.quote(part) for part in parts)


def _tmux_target_missing(stderr: str) -> bool:
    normalized = stderr.lower()
    return (
        "can't find pane" in normalized
        or "can't find session" in normalized
        or "no server running" in normalized
        or "failed to connect" in normalized
        or "server exited unexpectedly" in normalized
    )


__all__ = [
    "TmuxController",
    "TmuxControllerProtocol",
    "_tmux_target_missing",
]

File: dify-agent-sandbox-runtime/internal/server/tmux.go
package server

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// TmuxController manages tmux sessions for shellctl jobs via a dedicated socket.
type TmuxController struct {
	config *Config
}

// NewTmuxController creates a new tmux controller.
func NewTmuxController(config *Config) *TmuxController {
	return &TmuxController{config: config}
}

// StartServer ensures the tmux server is running.
func (t *TmuxController) StartServer() error {
	_, err := t.runTmux("start-server")
	return err
}

// ListSessions returns the set of active tmux session names.
func (t *TmuxController) ListSessions() (map[string]bool, error) {
	result, err := t.runTmuxNoCheck("list-sessions", "-F", "#{session_name}")
	if err != nil {
		return nil, err
	}
	if result.exitCode != 0 {
		stderr := strings.TrimSpace(result.stderr)
		if isTmuxTargetMissing(stderr) {
			return make(map[string]bool), nil
		}
		return nil, NewServerError(500, "tmux_error", stderr)
	}
	sessions := make(map[string]bool)
	for _, line := range strings.Split(strings.TrimSpace(result.stdout), "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			sessions[line] = true
		}
	}
	return sessions, nil
}

// SessionExists checks if a tmux session exists by name.
func (t *TmuxController) SessionExists(sessionName string) (bool, error) {
	sessions, err := t.ListSessions()
	if err != nil {
		return false, err
	}
	return sessions[sessionName], nil
}

// IsOutputPipeActive checks the #{pane_pipe} format variable for a job pane.
// Returns: true=active, false=inactive, error with nil bool if pane missing.
func (t *TmuxController) IsOutputPipeActive(jobID string) (*bool, error) {
	result, err := t.runTmuxNoCheck(
		"display-message", "-p", "-t", JobPaneTarget(jobID), "#{pane_pipe}",
	)
	if err != nil {
		return nil, err
	}
	if result.exitCode != 0 {
		stderr := strings.TrimSpace(result.stderr)
		if isTmuxTargetMissing(stderr) {
			return nil, nil // pane missing
		}
		return nil, NewServerError(500, "tmux_error", stderr)
	}
	active := strings.TrimSpace(result.stdout) == "1"
	return &active, nil
}

// CreateJobSession creates a new tmux session for a job.
func (t *TmuxController) CreateJobSession(jobID, jobDir, cwd string, cols, rows int) error {
	runnerCmd := shellJoin([]string{
		t.config.RunnerPath(), jobDir, jobID, cwd,
	})
	result, err := t.runTmuxNoCheck(
		"-f", "/dev/null",
		"new-session", "-d",
		"-s", JobSessionName(jobID),
		"-x", fmt.Sprintf("%d", cols),
		"-y", fmt.Sprintf("%d", rows),
		runnerCmd,
	)
	if err != nil {
		return err
	}
	if result.exitCode != 0 {
		return NewServerError(500, "tmux_new_session_failed",
			strings.TrimSpace(result.stderr))
	}
	return nil
}

// EnableOutputPipe attaches the sanitize→output pipeline via tmux pipe-pane.
func (t *TmuxController) EnableOutputPipe(jobID, jobDir string, readyFile string) error {
	pipeCmd := t.buildPipeCommand(jobID, jobDir, readyFile)
	result, err := t.runTmuxNoCheck(
		"pipe-pane", "-o", "-t", JobPaneTarget(jobID), pipeCmd,
	)
	if err != nil {
		return err
	}
	if result.exitCode != 0 {
		return NewServerError(500, "pipe_failed",
			strings.TrimSpace(result.stderr))
	}
	return nil
}

// SendInput sends text to a job's tmux pane via load-buffer + paste-buffer.
func (t *TmuxController) SendInput(jobID, text string) error {
	bufferName := fmt.Sprintf("shellctl-in-%s", jobID)

	// Write input to a temp file
	tmpFile, err := os.CreateTemp(t.config.RuntimeDir, fmt.Sprintf("shellctl-input-%s-", jobID))
	if err != nil {
		return err
	}
	tmpPath := tmpFile.Name()
	defer func() { _ = os.Remove(tmpPath) }()

	if _, err := tmpFile.WriteString(text); err != nil {
		_ = tmpFile.Close()
		return err
	}
	if err := tmpFile.Close(); err != nil {
		return err
	}

	// Load buffer
	result, err := t.runTmuxNoCheck("load-buffer", "-b", bufferName, tmpPath)
	if err != nil {
		return err
	}
	if result.exitCode != 0 {
		stderr := strings.TrimSpace(result.stderr)
		if isTmuxTargetMissing(stderr) {
			return NewServerError(409, "tmux_target_missing", stderr)
		}
		return NewServerError(500, "tmux_input_failed", stderr)
	}

	// Paste buffer
	result, err = t.runTmuxNoCheck(
		"paste-buffer", "-t", JobPaneTarget(jobID), "-b", bufferName,
	)
	// Always clean up buffer
	_, _ = t.runTmuxNoCheck("delete-buffer", "-b", bufferName)

	if err != nil {
		return err
	}
	if result.exitCode != 0 {
		stderr := strings.TrimSpace(result.stderr)
		if isTmuxTargetMissing(stderr) {
			return NewServerError(409, "tmux_target_missing", stderr)
		}
		return NewServerError(500, "tmux_input_failed", stderr)
	}
	return nil
}

// SendInterrupt sends C-c to a job's pane.
func (t *TmuxController) SendInterrupt(jobID string) error {
	_, _ = t.runTmuxNoCheck("send-keys", "-t", JobPaneTarget(jobID), "C-c")
	return nil
}

// CleanupSession kills the tmux session for a job (best-effort).
func (t *TmuxController) CleanupSession(jobID string) {
	_, _ = t.runTmuxNoCheck("kill-session", "-t", JobSessionName(jobID))
}

func (t *TmuxController) buildPipeCommand(jobID, jobDir, readyFile string) string {
	sanitizeCmd := shellJoin(append(t.config.SanitizePtyCommand, "--ready-file", readyFile))
	runnerExitCmd := shellJoin(append(t.config.RunnerExitCommand,
		"--state-dir", t.config.StateDir,
		"--job-id", jobID,
		"--sqlite-busy-timeout-ms", fmt.Sprintf("%d", t.config.SQLiteBusyTimeoutMs),
	))

	outputPath := shellQuote(filepath.Join(jobDir, "output.log"))
	drainedPath := shellQuote(filepath.Join(jobDir, ".pipe-drained"))
	errorLogPath := shellQuote(filepath.Join(jobDir, "pipe-error.log"))
	failedPath := shellQuote(filepath.Join(jobDir, ".pipe-failed"))
	exitCodePath := shellQuote(filepath.Join(jobDir, "runner-exit-code"))
	endedAtPath := shellQuote(filepath.Join(jobDir, "runner-ended-at"))

	parts := []string{
		fmt.Sprintf("%s >> %s 2> %s", sanitizeCmd, outputPath, errorLogPath),
		"sanitize_status=$?",
		"runner_exit_status=0",
		fmt.Sprintf(
			`if [ "$sanitize_status" -eq 0 ]; then : > %s; if [ -s %s ] && [ -s %s ]; then %s --exit-code "$(cat %s)" --ended-at "$(cat %s)" 2>> %s; runner_exit_status=$?; if [ "$runner_exit_status" -ne 0 ]; then printf 'runner-exit failed with status %%s\n' "$runner_exit_status" >> %s; fi; fi; else : > %s; fi`,
			drainedPath, exitCodePath, endedAtPath, runnerExitCmd,
			exitCodePath, endedAtPath, errorLogPath, errorLogPath, failedPath,
		),
		`if [ "$sanitize_status" -ne 0 ]; then exit "$sanitize_status"; fi`,
		`exit "$sanitize_status"`,
	}
	return strings.Join(parts, " ; ")
}

type tmuxResult struct {
	stdout   string
	stderr   string
	exitCode int
}

func (t *TmuxController) runTmux(args ...string) (string, error) {
	result, err := t.runTmuxNoCheck(args...)
	if err != nil {
		return "", err
	}
	if result.exitCode != 0 {
		stderr := strings.TrimSpace(result.stderr)
		if stderr == "" {
			stderr = "tmux command failed"
		}
		return "", NewServerError(500, "tmux_error", stderr)
	}
	return result.stdout, nil
}

func (t *TmuxController) runTmuxNoCheck(args ...string) (*tmuxResult, error) {
	cmdArgs := append([]string{"-S", t.config.TmuxSocket()}, args...)
	cmd := exec.Command("tmux", cmdArgs...)

	// Clear TMUX env to avoid nesting
	env := os.Environ()
	filtered := env[:0]
	for _, e := range env {
		if !strings.HasPrefix(e, "TMUX=") {
			filtered = append(filtered, e)
		}
	}
	cmd.Env = filtered

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			return nil, fmt.Errorf("exec tmux: %w", err)
		}
	}

	return &tmuxResult{
		stdout:   stdout.String(),
		stderr:   stderr.String(),
		exitCode: exitCode,
	}, nil
}

func isTmuxTargetMissing(stderr string) bool {
	lower := strings.ToLower(stderr)
	return strings.Contains(lower, "can't find pane") ||
		strings.Contains(lower, "can't find session") ||
		strings.Contains(lower, "no server running") ||
		strings.Contains(lower, "failed to connect") ||
		strings.Contains(lower, "server exited unexpectedly")
}

func shellJoin(parts []string) string {
	quoted := make([]string, len(parts))
	for i, p := range parts {
		quoted[i] = shellQuote(p)
	}
	return strings.Join(quoted, " ")
}

func shellQuote(s string) string {
	return "'" + strings.ReplaceAll(s, "'", `'\''`) + "'"
}
